hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/KeyLifecycleService.java

```diff
@@ -366,171 +371,302 @@ private void evaluateFSOBucket(OmVolumeArgs volume, OmBucketInfo bucket, String
+          }
+
+          if (!subDirSummary.getSubDirList().isEmpty()) {
+            item.setSubDirSummary(subDirSummary);
+            try {
+              stack.push(item);
+            } catch (CapacityFullException e) {
+              LOG.warn("Abort evaluate {}/{} at {}", volumeName, bucketName, currentDirPath, e);
+              return;
+            }
+
+            // depth first evaluation, push subDirs into stack
+            for (OmDirectoryInfo subDir : subDirSummary.getSubDirList()) {
+              String subDirPath = currentDirPath.isEmpty() ? subDir.getName() :
+                  currentDirPath + OM_KEY_PREFIX + subDir.getName();
+              try {
+                stack.push(new PendingEvaluateDirectory(subDir, subDirPath, null));
+              } catch (CapacityFullException e) {
+                LOG.warn("Abort evaluate {}/{} at {}", volumeName, bucketName, subDirPath, e);
+                return;
+              }
+            }
+            newSubDirPushed = true;
+          }
+        } else {
+          // this item is a parent directory, check how many sub directories are deleted.
+          for (OmDirectoryInfo subDir : subDirSummary.getSubDirList()) {
+            if (deletedDirSet.remove(subDir.getObjectID())) {
+              deletedDirCount++;
+            }
+          }
+        }
+
+        if (newSubDirPushed) {
+          continue;
+        }
+
+        // evaluate direct files, first check cache, then check table
+        // there are three cases:
+        // a. key is deleted in cache, while it's not deleted in table yet
+        // b. key is new added in cache, not in table yet
+        // c. key is updated in cache(rename), but not updated in table yet
+        //    in this case, the fromKey is a deleted key in cache, and the toKey is a newly added key in cache,
+        //    and fromKey is also in table
+        long numKeysUnderDir = 0;
+        long numKeysExpired = 0;
+        HashSet<String> deletedKeySet = new HashSet();
+        Iterator<Map.Entry<CacheKey<String>, CacheValue<OmKeyInfo>>> cacheIter = keyTable.cacheIterator();
+        while (cacheIter.hasNext()) {
+          Map.Entry<CacheKey<String>, CacheValue<OmKeyInfo>> entry = cacheIter.next();
+          OmKeyInfo key = entry.getValue().getCacheValue();
+          if (key == null) {
+            deletedKeySet.add(entry.getKey().getCacheKey());
+            continue;
+          }
+          numKeysUnderDir++;
+          String keyPath = currentDirPath.isEmpty() ? key.getKeyName() :
+              currentDirPath + OM_KEY_PREFIX + key.getKeyName();
+          for (OmLCRule rule : ruleList) {
+            if (key.getParentObjectID() == currentDirObjID && rule.match(key, keyPath)) {
+              // mark key as expired, check next key
+              if (keyList.isFull()) {
+                // if keyList is full, send delete/rename request for expired keys
+                handleAndClearFullList(bucket, keyList, false);
```

Comment:
The handled key in here may duplicate with the L599, A key may exist in both DB and cache within a certain period of time. maybe we can record the handled key in cache and skip it in DB iterator.

Reply:
Good point.